# Project 7 -- Kennedy Zapalac

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1
Consider the dataset `iowa_liquor_sales_cleaner.txt`.

In Project 05’s Question 2, you were asked to find out which Store Name had the most purchases, and once you obtained the store name, you calculated the total cost from that store. Now let’s use `awk associative arrays` to list each of the top 10 stores' total cost of purchases (per store).

(In Project 5 we used column 19, which was the `State Bottle Cost`. This time, instead, let’s focus on column 22.)

For this question, please use column 22, which is the `Sale (Dollars)`.

a. Find the total cost of the purchases for each store. For the output, show only the top 10 stores, in terms of the total cost of the purchases. List each of these top 10 stores and the total cost of the purchases of each.

In [6]:
%%bash
# for each of the stores (names stored in field 4), sum the total cost of purchases, then print the total cost of purchases for each store, sort (g
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{mytotal[$4] += $22} END{for(i in mytotal) {print(mytotal[i], i)}}' | sort -gr | head -n10

1.23313e+08 HY-VEE #3 / BDI / DES MOINES
1.08328e+08 CENTRAL CITY 2
4.87686e+07 HY-VEE WINE AND SPIRITS / IOWA CITY
4.13889e+07 SAM'S CLUB 8162 / CEDAR RAPIDS
3.96246e+07 SAM'S CLUB 6344 / WINDSOR HEIGHTS
3.42666e+07 BENZ DISTRIBUTING
3.38266e+07 WILKIE LIQUORS
3.09229e+07 LOT-A-SPIRITS
2.79315e+07 SAM'S CLUB 8238 / DAVENPORT
2.68169e+07 CENTRAL CITY LIQUOR, INC.


HY-VEE \#3 is the top store with the \$123,313,000 in purchases

## Question 2
Let’s look at the dataset `escaped2020sample.txt`.

a. This dataset contains the CITY(column 9), STATE (column 10), and TRANSACTION_AMT (column 15) for each donation. Please calculate the total transaction amounts for each city/state pair from this dataset (for instance, West Lafayette, IN). For your answer, list the top 10 city/state pairs with the largest total transaction amounts. Please use awk associative arrays for this question.

In [3]:
%%bash
# substitute the double quotes (") using sed, s stands for substitute, what I am replacing is after the first "/", and what I want to replace it with is after the second "/"
# sum the total transaction amount for each city, state pair, the print out the 10 city, state pairs with the greatest toal transaction amount
cat /anvil/projects/tdm/data/election/escaped2020sample.txt | sed 's/"//g' | awk -F"|" '{mytotal[$9","$10] += $15} END{for(i in mytotal) {print(mytotal[i], i)}}' | sort -nr | head -n10

331576505 NEW YORK,NY
38701698 SAN FRANCISCO,CA
37152718 WASHINGTON,DC
14141160 LOS ANGELES,CA
12463126 HOUSTON,TX
11642460 DALLAS,TX
11032607 CHICAGO,IL
8287423 BOSTON,MA
7141597 ATLANTA,GA
7115650 SEATTLE,WA


Ney York, NY has the largest total transaction amoutn of \$331,576,505

## Question 3
Now let us take a look at the dataset `1990.csv`.

a. You may have noticed that the "FlightDate" column (6th column) contains dates formatted as "1990-01-31". Please write an awk command to extract the year and month (not the day) from this column and then reformat them as (for instance) "01/1990". For your output, print each of the twelve months from 1990, and the number of flights that occur during each of those months.

In [15]:
%%bash
# grab the flight data column
# then split into 3 fields separated by "-", grab the year and month (if year = 1990, but this isn't necessary because all the flights are from 1990 but it conveniently removes the older header)
# sort by year and month, then count the number of flight for each year and month pair
cat /anvil/projects/tdm/data/flights/1990.csv | awk -F, '{print $6}' | awk -F"-" '{if($1 == "1990") {print $2"/"$1}}' | sort | uniq -c

 442700 01/1990
 396926 02/1990
 444320 03/1990
 433420 04/1990
 447371 05/1990
 437395 06/1990
 455057 07/1990
 457886 08/1990
 435638 09/1990
 455047 10/1990
 428781 11/1990
 436352 12/1990


August had the greatest number of flights in 1990

## Question 4
a. Use awk to create a new dataset from `1990.csv` called `1990_flight_info.csv`. This new file should include the following columns: the flight_month_year(MM/YYYY) just the same as you created in question 3, and also the total number of flights during that month_year. Order the results according to the number of flights in the month_year, from smallest to largest. 

The header of this file called 1990_flight_info.csv should look like:

flight_month_year;total_number_of_flights

In [4]:
%%bash
# same as above, but then remove the leading white space, separate the fields using a colon instead of space, add headers, and output to file
cat /anvil/projects/tdm/data/flights/1990.csv | awk -F, '{print $6}' | awk -F"-" '{if($1 == "1990") {print $2"/"$1}}' | sort | uniq -c | sed 's/^\s*//' | awk -F" " 'BEGIN{print "flight_month_year;total_number_of_flights"} {print $2";"$1}' > 1990_flight_info.csv

In [8]:
%%bash
# check the output of the new file I created
cat ~/1990_flight_info.csv

flight_month_year;total_number_of_flights
01/1990;442700
02/1990;396926
03/1990;444320
04/1990;433420
05/1990;447371
06/1990;437395
07/1990;455057
08/1990;457886
09/1990;435638
10/1990;455047
11/1990;428781
12/1990;436352


## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.